File: java/src/main/java/io/cucumber/prettyformatter/AmbiguousStepDefinitionsFormatter.java
package io.cucumber.prettyformatter;

import io.cucumber.messages.types.StepDefinition;

import java.util.List;

import static io.cucumber.prettyformatter.Theme.Element.LOCATION;

final class AmbiguousStepDefinitionsFormatter {

    private final int indentation;
    private final SourceReferenceFormatter sourceReferenceFormatter;
    private final Theme theme;

    private AmbiguousStepDefinitionsFormatter(Theme theme, int indentation, SourceReferenceFormatter sourceReferenceFormatter) {
        this.theme = theme;
        this.indentation = indentation;
        this.sourceReferenceFormatter = sourceReferenceFormatter;
    }

    static Builder builder(SourceReferenceFormatter sourceReferenceFormatter, Theme theme) {
        return new Builder(sourceReferenceFormatter, theme);
    }

    void formatTo(List<StepDefinition> stepDefinitions, LineBuilder lineBuilder) {
        lineBuilder.indent(indentation)
                .append("Multiple matching step definitions found:")
                .newLine();
        for (StepDefinition stepDefinition : stepDefinitions) {
            lineBuilder.indent(indentation)
                    .append("  ")
                    .append(theme.bulletPointIcon())
                    .append(" ")
                    .append(stepDefinition.getPattern().getSource());
            sourceReferenceFormatter.format(stepDefinition.getSourceReference())
                    .ifPresent(location -> lineBuilder.append(" ").append(LOCATION, "# " + location));
            lineBuilder.newLine();
        }
    }

    static final class Builder {
        private final SourceReferenceFormatter sourceReferenceFormatter;
        private final Theme theme;
        private int indentation = 0;

        private Builder(SourceReferenceFormatter sourceReferenceFormatter, Theme theme) {
            this.sourceReferenceFormatter = sourceReferenceFormatter;
            this.theme = theme;
        }

        Builder indentation(int indentation) {
            this.indentation = indentation;
            return this;
        }

        AmbiguousStepDefinitionsFormatter build() {
            return new AmbiguousStepDefinitionsFormatter(theme, indentation, sourceReferenceFormatter);
        }
    }
}

File: cpp/src/cucumber/pretty-formatter/AmbiguousStepDefinitionsFormatter.cpp
#include "cucumber/pretty-formatter/AmbiguousStepDefinitionsFormatter.hpp"
#include "cucumber/messages/StepDefinition.hpp"
#include "cucumber/pretty-formatter/LineBuilder.hpp"
#include "cucumber/pretty-formatter/SourceReferenceFormatter.hpp"
#include "cucumber/pretty-formatter/Theme.hpp"
#include <cstddef>
#include <memory>
#include <string>
#include <utility>
#include <vector>

namespace cucumber::pretty_formatter
{
    AmbiguousStepDefinitionsFormatter::AmbiguousStepDefinitionsFormatter(std::shared_ptr<Theme> theme,
        SourceReferenceFormatter sourceReferenceFormatter)
        : AmbiguousStepDefinitionsFormatter{ 0, std::move(theme), std::move(sourceReferenceFormatter) }
    {}

    AmbiguousStepDefinitionsFormatter::AmbiguousStepDefinitionsFormatter(std::size_t indent, std::shared_ptr<Theme> theme,
        SourceReferenceFormatter sourceReferenceFormatter)
        : indent{ indent }
        , theme{ std::move(theme) }
        , sourceReferenceFormatter{ std::move(sourceReferenceFormatter) }
    {}

    void AmbiguousStepDefinitionsFormatter::Format(LineBuilder& lineBuilder,
        const std::vector<std::shared_ptr<const messages::StepDefinition>>& stepDefinitions)
    {
        lineBuilder.Indent(indent).Append("Multiple matching step definitions found:").NewLine();

        for (const auto& stepDefinition : stepDefinitions)
        {
            lineBuilder.Indent(indent).Append("  ").Append(theme->BulletPointIcon()).Append(" ").Append(stepDefinition->pattern->source);

            const auto optUri = sourceReferenceFormatter.Format(stepDefinition->sourceReference);
            if (optUri.has_value())
            {
                lineBuilder.Append(" ").Append(Theme::Element::location, "# " + optUri.value());
            }
            lineBuilder.NewLine();
        }
    }
}
